modules/core/src/shadertools/src/modules/picking/picking.js

```diff
@@ -77,7 +77,18 @@ const float COLOR_SCALE = 1. / 255.;
  */
 vec4 picking_filterHighlightColor(vec4 color) {
   bool selected = bool(picking_vRGBcolor_Aselected.a);
-  return selected ? (picking_uHighlightColor * COLOR_SCALE) : color;
+
+  if (selected) {
+    vec4 highLightColor = picking_uHighlightColor * COLOR_SCALE;
+
+    float resultAlpha = highLightColor.a + color.a * (1.0 - highLightColor.a);
+
+    vec3 resultRGB = (highLightColor.rgb * highLightColor.a
+     + color.rgb * color.a * (1.0 - highLightColor.a)) / resultAlpha;
+    return vec4(resultRGB, color.a);
```

Comment:
return vec4(resultRGB, resultAlpha);

Reply:
`color.a` actually works better. In one use case we create arrows along a path by manipulating alpha:

![image](https://user-images.githubusercontent.com/2059298/49543281-a3156b80-f88c-11e8-8e79-652657ec0222.png)

The hollow part is still pickable, and on highlight it should not be filled.

Reply:
Good point, then better to change vec3 resultRGB = (highLightColor.rgb * highLightColor.a + color.rgb * color.a * (1.0 - highLightColor.a)) / resultAlpha;
to 
vec3 resultRGB = (highLightColor.rgb * highLightColor.a + color.rgb * color.a * (1.0 - highLightColor.a)) / color.a; and handle color.a case

Reply:
Mathematically this commit is correct. The way it is written is very confusing. It is equivalent to:

```glsl
float highLightAlpha = highLightColor.a;
float highLightRatio = highLightAlpha / (highLightAlpha + color.a * (1.0 - highLightAlpha));
vec3 resultRGB = mix(color.rgb, highLightColor.rgb, blendRatio);
```

Reply:
To be honest I think this is totally fine too:

```glsl
vec3 resultRGB = mix(color.rgb, highLightColor.rgb, highLightColor.a);
```

Reply:
> vec3 resultRGB = mix(color.rgb, highLightColor.rgb, highLightColor.a)

Sounds good.

On whether to use actual alpha or highlight alpha, each has its side effects. When using highlight.alpha we might end up making things opaque that are supposed to transparent (like arrow), but when using actual alpha, and if it is very low, you might not notice it is highlighted (defeats purpose of highlighting). At least using highlight.alpha you can change the behavior by changing highlight color prop, but when using actual alpha you don't have that control. I would still vote for hightlight.alpha :) , but using actual alpha is not a blocker.

Reply:
That's why better to calculate resultRGB by dividing color.a, which can guarantee hightlight color's RGB component is preserved in final rendering color. Since our rendering system is not based on premultiplied alpha, need handle divided by 0 case with this solution, which is not ideal.